### ISSUE TITLE
Web Api - Concurrency, Read & Writes

### ISSUE THREAD

**Issue by bobmcallan** [mbdavid/LiteDB#736]
Hello, been using LiteDb since v2, and is an excellent repo. Recently I'm importing larger datasets and have encountered some concurrency issues, which is expected however I'd like to understand the best approach to solve.

I've seen discussions regarding 2 options, however which is the preferred design.

1. (current) Open and close as required @ a basic read/write/query level.
2. Wrap the entire API with a static/Lazy "new LiteDatabase(...)" and have all functions use this. 

I currently have the a wrapper around most LiteDb functions and use a base class to provide connection and naming details. This is effective, however as stated above there are some locking errors which inserting large (+5k) data sets.

I'm current running v3. I have tried v4, however there is a bug in the Nuget package for inserts (returns void).

Thanks again,
Bob

**Comment by mbdavid** [comment]
Hi @bobmcallan, v3 has an issue when using concurrency access. To fix this, I re-write a lot of code and was called v4.
This missing insert return was fixed in LiteDB master branch. I will submit new nuget version this weekend.

**Comment by bobmcallan** [comment]
I've updated to 4.0.1. All is working as required. 

I did discover one issue when serialising a list to Json. Maynot be an issue, as more my poor coding  practice. 

**Issue**: If the code does not convert to list with-in the 'using' statement a post json serialisation will fail.

**Errors** : var litedbTable = collection.FindAll().Take(100);
**Passes**: var litedbTable = collection.FindAll().Take(100).ToList();

***********************************************
CODE
***********************************************
            using (var db = new LiteDatabase(_connectionString))
            {
                if (!db.CollectionExists(name))
                {
                    return Ok("Nope");
                }

                try
                {
                    var collection = db.GetCollection(name);
                    // var litedbTable = collection.FindAll().Take(100).ToList();
                    var litedbTable = collection.FindAll().Take(100);

                    if (litedbTable != null)
                        return Ok(litedbTable.toLiteDbTable());
                    else
                        return Ok("Nope");
                }
                catch (Exception ex)
                {
                    Log.Error(ex, "Not now!");
                    return Content(HttpStatusCode.InternalServerError, ex);
                }
            }
***********************************************

***********************************************
TRACE
***********************************************
Message: "An error has occurred.",
ExceptionMessage: "Object reference not set to an instance of an object.",
ExceptionType: "System.NullReferenceException",
StackTrace: "   at LiteDB.StreamExtensions.<>c__DisplayClass4_0.<TryLock>b__0()
   at LiteDB.FileHelper.TryExec(Action action, TimeSpan timeout)
   at LiteDB.StreamExtensions.TryLock(FileStream stream, Int64 position, Int64 length, TimeSpan timeout)
   at LiteDB.FileDiskService.Lock(LockState state, TimeSpan timeout)
   at LiteDB.LockService.Read()
   at LiteDB.LiteEngine.<Find>d__9.MoveNext()
   at LiteDB.LiteEngine.<Find>d__13.MoveNext()
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()
   at System.Linq.Enumerable.<TakeIterator>d__24`1.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at System.Net.Http.Formatting.BaseJsonMediaTypeFormatter.WriteToStream(Type type, Object value, Stream writeStream, Encoding effectiveEncoding)
   at System.Net.Http.Formatting.JsonMediaTypeFormatter.WriteToStream(Type type, Object value, Stream writeStream, Encoding effectiveEncoding)
   at System.Net.Http.Formatting.BaseJsonMediaTypeFormatter.WriteToStreamAsync(Type type, Object value, Stream writeStream, HttpContent content, TransportContext transportContext, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Owin.HttpMessageHandlerAdapter.<BufferResponseContentAsync>d__13.MoveNext()"

**Comment by mbdavid** [comment]
Hi @bobmcallan, when you use `IEnumerable` the data access occurs only when you use this result (not when you Find method). In this case, serialization will use this but only after already closed datafile. That why you need use ToList before, to load all to memory.
